### PR TITLE
CT Surface Sectioned Geometry

### LINKED ISSUE #224: Surface Sectioned Geometry Template
## Description & Tasks
Geometry template for the definition of sectioned surfaces 

- [ ] Author Surface Sectioned Geometry  Template
- [ ] Documentation of Body Surface Sectioned Geometry  Template
- [ ]  Add RepresentationType, if Needed

### Discussion Points

1. Does this require an additional RepresentationType or is it going to sit under Surface3D?

### Links
[Miro Element](https://miro.com/app/board/o9J_l5rJCLk=/?moveToWidget=3074457362582256473&cot=14)
[Miro Task Card](https://miro.com/app/board/o9J_l5rJCLk=/?moveToWidget=3074457362911087571&cot=14)

### Decisions 
TBD

### PR DESCRIPTION
**Pull request by larswik**
Proposal for new CT "Surface Sectioned Geometry". Addresses #224
In this process I also propose a new RepresentationIdentifier = "SectionedSurface". A little bit "unbalanced" vs the corresponding for sectioned solids which are defined using "AdvancedSweptSolid", a level missing for surfaces. Since Surface Sectioned Geometry was proposed to be placed as a sibling to "Surface3D geometry"/RepresentationIdentifier="Surface3D" and "Surface Tesselation Geometry" /RepresentationIdentifier="Tesselation", I figured that this would be my proposal.

### CODE REVIEW

**Review by jmirtsch — CHANGES_REQUESTED** [review]
Fix UniqueIds

**Comment by SergejMuhic** [comment]
The proposed changes require a schema change and documentation change also.

**Comment by larswik** [comment]
Besides the documentation change, is it IfcShapeRepresentationTypes that you are referring to? I felt that the way things are structured for RepresentationType, it felt to me as if something new was needed (maybe even ideally an "AdvancedSweptSurface"). However if it is ok to just use "SurfaceModel", then fine by me. I am not entirely sure how these are actually used in software. Happy to remove that part of the proposal.

**Comment by SergejMuhic** [comment]
Yes, that is what I am referring to. Adding 'SectionedSurface' means we would modify the function.

File: IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+															<Rules>
+																<DocModelRuleEntity Name="IfcBoundedCurve" />
+															</Rules>
+														</DocModelRuleAttribute>
+														<DocModelRuleAttribute Name="CrossSections" Identification="CrossSections">
+															<Rules>
+																<DocModelRuleEntity Name="IfcProfileDef" />
+															</Rules>
+														</DocModelRuleAttribute>
+														<DocModelRuleAttribute Name="CrossSectionPositions" Identification="CrossSectionPositions">
```
Comment (SergejMuhic) [inline IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:56]: This I would remove. Same as Sectioned solid. It is restricted in the template already.

**Comment by SergejMuhic** [comment]
> 
> 
> Besides the documentation change, is it IfcShapeRepresentationTypes that you are referring to? I felt that the way things are structured for RepresentationType, it felt to me as if something new was needed (maybe even ideally an "AdvancedSweptSurface"). However if it is ok to just use "SurfaceModel", then fine by me. I am not entirely sure how these are actually used in software. Happy to remove that part of the proposal.

The purpose of the identifier and type has somewhat diminished with mvdxml. Not sure if we are ready to remove it completely (especially now) since it does provide a mechanism to quite quickly identify whether the implementation supports it. This kind of also implies the necessity of "SectionedSurface" as type since `IfcSectionedSurface` is not your typical surface representation. Let me know what you want to do.

**Comment by larswik** [comment]
Removed CrossSectionPositions. Remains the issue about RepresentationType.

File: IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+												<DocModelRuleEntity Name="IfcGeometricRepresentationContext" />
+											</Rules>
+										</DocModelRuleAttribute>
+										<DocModelRuleAttribute CardinalityMin="1" CardinalityMax="1" Name="RepresentationIdentifier" Identification="Identifier">
+											<Rules>
+												<DocModelRuleEntity Name="IfcLabel">
+													<Rules>
+														<DocModelRuleAttribute Name="Value=Surface" Description="Value=Surface" />
+													</Rules>
+												</DocModelRuleEntity>
```
Comment (AlexBrad1eyCT) [inline IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:23]: This is the wrong encoding it should look the same as the RepresentationType Below, this needs correcting

**Review by AlexBrad1eyCT — CHANGES_REQUESTED** [review]
Only the error on the inherited RepresentationIdentifier attribute needs correcting.

File: IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml (hunk 2 above)
Comment (larswik) [reply IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:23]: Seems like there is an error in the encoding at the level "Surface Geometry", right? Didn't touch that but will look into it.

Comment (AlexBrad1eyCT) [reply IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:23]: yes if you merge the develop you should get the correct reference on Surface Geometry that just needs adding like in sectionedsolidHorizontal

Comment (AlexBrad1eyCT) [reply IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:23]:
```suggestion
												<DocModelRuleEntity Name="IfcLabel">
													<Rules>
														<DocModelRuleEntity xsi:nil="true" href="IfcLabel_212pjtt2TCSe6IuxauYKd"/>
													</Rules>
												</DocModelRuleEntity>
```

**Comment by larswik** [comment]
Out of my comfort zone here but this still looks wrong (Surface Geometry):
image
Shouldn't it be more in the format (Body Geometry):
image
However the cml encoding looks very similar...

**Review by AlexBrad1eyCT — APPROVED** [review]
LFTM

**Comment by AlexBrad1eyCT** [comment]
I saw that and compared the XML and its exactly the same (minus the ids and the literal ofcourse) so i cannot explain it.
some thing for the backlog i think @jmirtsch @SergejMuhic

**Comment by larswik** [comment]
I've made the change and the xml encoding looks ok. However, still missing the unique id generation so haven't pushed yet. I am in the Development branch of IfcDoc. Thought that was the correct one.

**Comment by AlexBrad1eyCT** [comment]
@larswik the xml is all correct the uniqueId attribute is only on the original rule so that is fine its just the display in IfcDoc seems to have a periodic error.

**Comment by AlexBrad1eyCT** [comment]
@larswik you must not be running the latest build, I just opened and saved your branch for this pull request and it did not change the sectioned surface rule. double check you have the latest development of IfcDoc and that it has been rebuilt the fix was only merged (PR87) on monday.

**Comment by larswik** [comment]
@AlexBrad1eyCT could you try to change the RepresentationIdentifier part of the Sectioned surface rule? It does not seem right at the moment. I have fetched and pulled everything from IfcDoc/Development.

File: IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml (hunk 1 above)
Comment (AlexBrad1eyCT) [reply IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml]: have been removed and resolved

File: IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+												<DocModelRuleEntity Name="IfcGeometricRepresentationContext" />
+											</Rules>
+										</DocModelRuleAttribute>
+										<DocModelRuleAttribute CardinalityMin="1" CardinalityMax="1" Name="RepresentationIdentifier" Identification="Identifier">
+											<Rules>
+												<DocModelRuleEntity Name="IfcLabel">
+													<Rules>
+														<DocModelRuleEntity xsi:nil="true" href="IfcLabel_212pjtt2TCSe6IuxauYKd"/>
+													</Rules>
+												</DocModelRuleEntity>
```
Comment (AlexBrad1eyCT) [inline IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:23]: Found the Issue Double bracketing should be fine now @larswik
```suggestion
												<DocModelRuleEntity xsi:nil="true" href="IfcLabel_212pjtt2TCSe6IuxauYKd"/>>
```

Comment (AlexBrad1eyCT) [reply IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/DocTemplateDefinition.xml:23]: image

**Comment by AlexBrad1eyCT** [comment]
@larswik please review and submit my fix.
@jmirtsch @SergejMuhic @pjanck please review and approve where necessary ASAP so we can get this PR closed and be able deliver the CTs

**Comment by larswik** [comment]
Thanks @AlexBrad1eyCT ! 
This LFTM now.

**Review by AlexBrad1eyCT — APPROVED** [review]
LFTM please review everyone!

**Review by jmirtsch — COMMENTED** [review]
It looks fine to me.

File: IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/Documentation.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+The _Surface Sectioned Geometry_ is the representation for elements having a having a 'Surface' representation where the surface is specified using two or more open profiles that are swept between specified positions along the directrix, e.g. for defining the upper boundary of an _IfcCourse_.
```
Comment (pjanck) [inline IFC4x3/Templates/Product Shape/Product Geometric Representation/Surface Geometry/Surface Sectioned Geometry/Documentation.md:1]:
```suggestion
The _Surface Sectioned Geometry_ is the representation for elements having a 'Surface' representation, where the surface is specified using two or more open profiles that are swept between specified positions along the directrix, e.g. for defining the upper boundary of an _IfcCourse_.
```

**Review by pjanck — CHANGES_REQUESTED** [review]
Small typo in the documentation, otherwise LFTM

**Review by AlexBrad1eyCT — APPROVED** [review]

**Pull request merged**